### ISSUE TITLE
"Unique Names" as column-headers in CSV reports

### ISSUE THREAD

**Issue by sbodorkos** [CIRDLES/Squid#333]
I'm still gathering threads on 204 overcount-corrections, Sample-specific common Pb corrections, etc.

In the meantime, could we revisit the idea of instituting column-headers corresponding to the superset of Unique Expression-Names defined by @NicoleRayner and I (see column L of [ExpressionNames.AcrossPerms_v5_NR+SB.xlsx](https://github.com/CIRDLES/Squid/files/3155904/ExpressionNames.AcrossPerms_v5_NR%2BSB.xlsx)), at least for the Reference Materials to begin with? I see this as an ongoing fundamental requirement of the reporting (at least for the CSV-based stuff): any solution we devise for expression-name 'aliasing' (and the provision of a smaller number of more easily-consumed expression-names for use in the Expression Manager) should be separate and additional to comprehensive reporting against unique expression-names.

I think it might also be useful for the unique-name CSV reporting to span the full gamut of unique built-in expression-names (irrespective of Perm-number, index isotope, or choice of Parent element as Concentration reference material). The resulting array would certainly be sparse for most Reference Material Reports, but I don't see that as a drawback, and a consistent structure would facilitate anything we build in order to enable 'easy' comparison of SQUID 2.50 against Squid3 outputs, for nominally equivalent data-reductions.

I have to confess that I have not yet thought very much about the SampleData sheet in SQUID 2.50, or the extent to which its apparently simple structure actually reflects the fact that the user has committed to exactly **one** of the 72 possible permutations (of Perm-number, index isotope, and Parent element for Concentration reference material). But the consistent number of columns in SQUID 2.50 SampleData leads me to believe that the Perm-dependent behaviour in the Unknowns Report Table **must** be simpler (in unique-name terms) than the Reference Material Report Tables... I guess we'll find out in due course.

**Comment by bowring** [comment]
Good idea!

**Question:** Will the new column headers then be rendered in a single row (cell in Excel-speak)?  This is something @NicoleRayner has requested as an option at least. 

One of the longest names is '4cor_8cor_PrimaryCalibConstDelta_Pct'.  (36 chars), which is currently rendered in the report using 4 lines (cells): "204corr", "8-corr", "206Pb/238U", "const delta%".

For readability, we could split names on the underscores to create multiple lines (cells) or not, depending on a user choice.  Alternatively, we could provide both solutions (one line, multi-line) in the CSV to minimize post-processing for various uses.

We could also give the user a mapping function with a nice interface to map Squid3 names to their own custom names, with the default range being the Squid3 names.  This functionality would be the first part of the planned feature to allow customization of the reports in general.

Thoughts?

**Comment by sbodorkos** [comment]
Yeah, I wondered about this, and then I decided to just leave it and hope you wouldn't ask me...

I do agree with @NicoleRayner that "single-row" column-headers are desirable, although it does limit the presentability of the CSVs. All I can suggest as partial remediation is that we replace all underscores with white spaces for delivery in the CSV, for readability and to hopefully facilitate Excel-style text-wrapping features.

I don't mind if we retain the multi-line column-headers as well (with "line breaks" matched to underscores in the unique names?), but I would like the "single-row" header to be **beneath** any multi-row rendition we retain (although obviously above the actual data).

Finally, for the comprehensive reporting, I think we could dispense with the sub-categorisation currently instituted in the top row of each reference Material Report CSV, and simply present the unique column-headers alphabetically. I see the SQUID 2.50-style sub-categorisation ("Correction-independent", "204Pb-corrected", etc.) as being more useful for a trimmed down CSV Report that takes full advantage of "knowing" the selected index isotope in order to deliver as "un-sparse" an array of data as possible (quite like SQUID 2.50), with sub-categorisation used to make the column-headers as simple as they can possibly be. 

But this latter functionality is probably contingent on how we handle the equivalent situation (true 'aliasing') in the Expression Manager, which is a job for another day.

**Comment by bowring** [comment]
I believe these issues have been addressed with the current report builder functionality.  If not, let's make new issue to deal with changes that might be needed.
